Feat/article/monitoring at scale with victoriametrics (#379)
BedrockStreaming/tech.bedrockstreaming.com: _posts/2022-09-06-monitoring-at-scale-with-victoriametrics.md

---
layout: post

> title: Monitoring at scale with victoria metrics

Reviewer: ```suggestion
title: Monitoring at scale with Victoria Metrics
```

description: why and how we moved from Prometheus to Victoria Metrics
author: j_menan
tags: [k8s, kubernetes, monitoring, prometheus, scaling, victoriametrics, cardinality]
color: rgb(251,87,66) # this is Bedrock color here
---

# Monitoring at BedRock :
At [Bedrock Streaming](https://www.bedrockstreaming.com/), a large part of our applications are hosted on Kubernetes clusters, others use the EC2 service from AWS and a small part are hosted on "OnPremise" servers.

> From 2018 until January 2022, we used Prometheus to monitor all these platforms, because Prometheus met all our needs: keeping control over our monitoring solution and supporting service discovery, which is essential on environments such as Kubernetes or AWS EC2. Prometheus can also support custom exporters that we developed internally.

Reviewer 2: ```suggestion
From 2018 until January 2022, we used Prometheus to monitor all these platforms, because Prometheus met all our needs: keeping control over our monitoring solution and supporting service discovery, which is essential in environments such as Kubernetes or AWS EC2. Prometheus can also support custom exporters that we developed internally.
```

# Limits of Prometheus
Prometheus does not have a native High Availability mode: to have high availability, we had to duplicate our Prometheus instances. This implies that our targets were "scrapped" by all our Prometheus instances (same for our rules and records).
To avoid this, we had to use sharding, but this made the infrastructure more complex. More information on this subject in this [documentation](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/high-availability.md) from the Prometheus operator

Prometheus is not designed to store metrics on a long-term basis, as mentioned in the [documentation](https://prometheus.io/docs/prometheus/latest/storage/#operational-aspects) : 

_> Prometheus's local storage is not intended to be durable long-term storage; external solutions offer extended retention and data durability._

Prometheus's local storage is not intended to be durable long-term storage; external solutions offer extended retention and data durability.
We worked around this limitation by using Victoria Metrics (VMCluster) as a LongTermStorage via the remote_write protocol

> All processes (scrapping, ingest, storage etc.) were, until now, managed in the same "prometheus" instance, which implied a less flexible and vertical scaling only (since recently a [Prometheus agent](https://prometheus.io/blog/2021/11/16/agent/) is available for the "scrapping" part).

Reviewer 2: ```suggestion
All processes (scrapping, ingest, storage, etc.) were, until now, managed in the same "prometheus" instance, which implied a less flexible and vertical scaling only (since recently a [Prometheus agent](https://prometheus.io/blog/2021/11/16/agent/) is available for the "scrapping" part).
```

The RAM and CPU usage of a Prometheus instance is correlated to the number of metrics (and their cardinality) it has to manage. In our case, several Prometheus instances 

> consumed more than 64 GB of RAM and 26 CPUs each, in order to absorb our peak loads. In a Kubernetes cluster, this high resources consumption can cause problems, especially for scheduling.

Reviewer: shouldn't this be right after the line before?

> The Write-Ahead Log (WAL) system can cause rather slow restarts if the Prometheus instance runs out of RAM and can cause the Prometheus instance to hang for varying lengths of time. During the replay of the WAL, prometheus doesn't scrape anything, thus there is no alerting and no way of knowing if something is going on

Reviewer: ```suggestion
The Write-Ahead Log (WAL) system can cause rather slow restarts if the Prometheus instance runs out of RAM and can cause the Prometheus instance to hang for varying lengths of time. During the replay of the WAL, Prometheus doesn't scrape anything, thus there is no alerting and no way of knowing if something is going on
```

## The cardinality of metrics
When our Kubernetes clusters manage a large number of pods, a constraint quickly appears: cardinality.

> _>The cardinality of a metric is the number of TimeSeries of that metric with single-valued labels._

Reviewer: ```suggestion
_> The cardinality of a metric is the number of TimeSeries of that metric with single-valued labels._
```

![schema of cardinality](/images/posts/2022-09-06-monitoring-at-scale-with-victoriametrics/cardinality-example.png)

In the example above, the `status_code` label has a cardinality of 5, app has a cardinality of 2 and the overall cardinality of the `server_reponses` metric is 10.

In this example, any Prometheus instance can handle this cardinality, but if you add for example the label `pod_name` or `client_IP` (or both) to the `server_reponses` metric, the cardinality increases for each different clients calls and for each pod.

You should read the excellent [article](https://www.robustperception.io/cardinality-is-key/) from "Robust Perception" for more details on this subject.

> At BedRock the high cardinality metrics comes from our HAProxy ingress. For our needs, we retrieve several labels like the name of the ingress pod as well as its IP address, but more importantly the name and IP address of the destination pod. In a cluster that can grow to  more than 15,000 pods, the combination of unique labels (cardinality) is very significant for some of our ingress metrics.

Reviewer 2: ```suggestion
At BedRock the high cardinality metrics come from our HAProxy ingress. For our needs, we retrieve several labels like the name of the ingress pod as well as its IP address, but more importantly the name and IP address of the destination pod. In a cluster that can grow to more than 15,000 pods, the combination of unique labels (cardinality) is very significant for some of our ingress metrics.
```

We found that Prometheus performed poorly when we had multiple metrics with high cardinalities (> 100,000), and resulted in over-consumption of RAM.

During a high load event, Prometheus could consume up to 200 GB of RAM before being OOMKilled. When this happened, we would go completely blind as we had no metrics or alerting.
This also impacts us on scalability in our Kubernetes clusters, as we use [CustomMetrics](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#scaling-on-custom-metrics) very heavily in HPAs to scale the number of pods in our applications.

_RAM and CPU consumption of our prometheus instances (the red lines represent the reboots of our instances, we also see a loss of metrics)_
![RAM consumption of prometheus ](/images/posts/2022-09-06-monitoring-at-scale-with-victoriametrics/prometheus-ram.png)
![RAM consumption of prometheus ](/images/posts/2022-09-06-monitoring-at-scale-with-victoriametrics/prometheus-cpu.png)

> Prometheus is still a good solution, which has served its purpose well for several years, but for our use case, prometheus has reached its limits in our production environments.

Reviewer: ```suggestion
Prometheus is still a good solution, which has served its purpose well for several years, but we have reached its limits in our production environments.
```
remove prometheus word repetition

 - [Remote Write Storage Wars](https://promcon.io/2019-munich/talks/remote-write-storage-wars/)
 - [Prometheus vs VictoriaMetrics benchmark on node_exporter metrics](https://valyala.medium.com/prometheus-vs-victoriametrics-benchmark-on-node-exporter-metrics-4ca29c75590f)
 - [When size matters — benchmarking VictoriaMetrics vs Timescale and InfluxDB](https://valyala.medium.com/when-size-matters-benchmarking-victoriametrics-vs-timescale-and-influxdb-6035811952d4)
 - [Comparing Thanos to VictoriaMetrics cluster](https://medium.com/faun/comparing-thanos-to-victoriametrics-cluster-b193bea1683)

We also wanted to keep the Prometheus language: [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) in order to keep our Grafana dashboards and all our Prometheus alerts. Even though Victoria Metrics offers its own MetricsQL language, it is perfectly compatible with PromQL.

You can see the [main features](https://docs.victoriametrics.com/#prominent-features) of Victoria Metrics as well as various [case studies](https://docs.victoriametrics.com/#case-studies-and-talks) in their documentation.

# POC of Victoria Metrics

> We wanted to validate the performance and consumption of a stack entirely based on Victoria Metrics, the results were really encouraging

Reviewer: ```suggestion
We wanted to validate the performance and consumption of a stack entirely based on Victoria Metrics, the results were really encouraging.
```

Alerting is also managed via a VMAlert component, which will execute the alerting and recording rules set by VictoriaMetrics. Notifications are managed by an Alertmanager which is also deployable via this chart.

One of the advantages of using this Helm chart is that it will deploy essential components to properly monitor a Kubernetes cluster such as _Kube-state-metrics_ or _prometheus-node-exporter_, but also scraping configurations for services such as _Kubelet, KubeApiServer, KubeControllerManager, KubeDNS, KubeEtcd, KubeScheduler, KubeProxy_

Alerting is also managed via a VMAlert component, which will execute the alerting and recording rules set by VictoriaMetrics. Notifications are managed by an Alertmanager which is also deployable via this chart.

_This is what our monitoring and alerting stack based on this Helm chart looks like._
![stack-vm](/images/posts/2022-09-06-monitoring-at-scale-with-victoriametrics/stack-vm.png)

# Resumption of the history

> We wanted to keep historical metrics of our Kubernetes clusters. Victoria Metrics provides a tool to manage the export and import of data from different tsdb: [vmctl](https://docs.victoriametrics.com/vmctl.html).

Reviewer: ```suggestion
We wanted to keep historical metrics of our Kubernetes clusters. Victoria Metrics provides a tool to manage the export and import of data from different TSDB: [vmctl](https://docs.victoriametrics.com/vmctl.html).
```
